spec.md

```diff
@@ -3767,6 +3912,282 @@ x.remove(2)                             # None (x == [1, 3])
 x.remove(2)                             # error: element not found
 ```
 
+<a id='set·add'></a>
+
+### set·add
+
+`S.add(x)` adds the value `x` to the set `S`. It returns `None`.
+
+It is permissible to `add` a value already present in the set; this leaves the
+set `S` unchanged.
+
+`add` fails if the set `S` is frozen or has active iterators, or if `x` is
+unhashable.
+
+If you need to add multiple elements to a set, see [`update`](#set·update) or
+the [`|=`](#sets) augmented assignment operation.
+
+<a id='set·clear'></a>
+
+### set·clear
+
+`S.clear()` removes all elements from the set `S`. It returns `None`.
+
+`clear` fails if the set `S` is frozen or has active iterators.
+
+<a id='set·difference'></a>
+
+### set·difference
+
+`S.difference(*others)` returns a new set containing elements found in the set
+`S` but not found in any of the iterable sequences `*others`.
+
+If `s` and `t` are sets, `s.difference(t)` is equivalent to `s - t`; however,
+note that the `-` operation requires both sides to be sets, while the
+`difference` method accepts arbitrary iterable sequences.
+
+It is permissible to call `difference` without any arguments; this returns a
+copy of the set `S`.
+
+```python
+set([1, 2, 3]).difference([2])             # set([1, 3])
+set([1, 2, 3]).difference([0, 1], [3, 4])  # set([2])
+```
+
+<a id='set·difference_update'></a>
+
+### set·difference\_update
+
+`S.difference_update(*others)` removes from the set `S` any elements found in
+any of the iterable sequences `*others`. It returns `None`.
+
+If `s` and `t` are sets, `s.difference_update(t)` is equivalent to `s -= t`;
+however, note that the `-=` augmented assignment requires both sides to be sets,
+while the `difference_update` method accepts arbitrary iterable sequences.
+
+It is permissible to call `difference_update` without any arguments; this leaves
+the set `S` unchanged.
+
+`difference_update` fails if the set `S` is frozen or has active iterators.
+
+```python
+s = set([1, 2, 3, 4])
+s.difference_update([2])             # None; s is set([1, 3, 4])
+s.difference_update([0, 1], [4, 5])  # None; s is set([3])
+```
+
+<a id='set·discard'></a>
+
+### set·discard
+
+`S.discard(x)` removes the value `x` from the set `S` if present. It returns
+`None`.
+
+It is permissible to `discard` a value not present in the set; this leaves the
+set `S` unchanged. If you want to fail on an attempt to remove a non-present
+element, use [`remove`](#set·remove) instead. If you need to remove multiple
+elements from a set, see [`difference_update`](#set·difference_update) or the
+[`-=`](#sets) augmented assignment operation.
+
+`discard` fails if the set `S` is frozen or has active iterators, even if `x` is
+not a member of the set.
```

Comment:
"`discard` fails if the set `S` is frozen or has active iterators, or if `x` is unhashable`. This applies even if `x` is not a member of the set."

(Slight grammar change to make it more likely to be read that the "x not a member" qualifier applies to all cases.)

Reply:
I disagree about failing if `x` is unhashable - an unhashable value cannot possibly be an element of a set, so no-oping is the more logical behavior. (It's also what Python 3 does.)

But I fixed the grammar as you suggested.

Reply:
> (It's also what Python 3 does.)

I don't think so.

```
>>> {1}.discard([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

That's in Python 3.11.9.

Reply:
Interesting! We found a place where python3 is internally inconsistent: you can discard `set` values (which in python3 are unhashable!) but not list or dict values.

```
>>> {1}.discard(set())
>>> {1}.discard({1})
>>> {1}.discard([1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
```

I think that this is might be a bug in python's `discard` implementation, since in other places - like in `isdisjoint` which you mention below - python3 does check for hashability.

Probably we should update the spec for consistency with python's platonic ideal - i.e. check for hashability everywhere.

Reply:
Fixed - added requirement for the argument to discard (and all other set methods) to not contain unhashable values.